src/ledger/test/LedgerCloseMetaStreamTests.cpp

```diff
@@ -181,6 +220,10 @@ TEST_CASE("LedgerCloseMetaStream file descriptor - REPLAY_IN_MEMORY",
 
     // Step 4: reopen the file as an XDR stream and read back the LCMs
     // and check they have the expected content.
+    //
+    // The !UNSAFE_EMIT_META case should still have streamed the latest meta,
+    // because catchup should have validated that ledger's hash against the
+    // history archive's version of it.
```

Comment:
Change comment: we don't validate against history archive, we validate history archive contents against user-provided or network-observed trusted hash.

It is also not clear to me that in OFFLINE_COMPLETE catchup, we will actually have streamed the last meta. I don't think we will, given my reading of the code. Not that it matters to users who are passing in `--start-ledger` and `--start-hash`, but in this _particular_ test, I think it should be failing when `unsafeEmitMeta==false`. Maybe double check?

Reply:
> Change comment: we don't validate against history archive, we validate history archive contents against user-provided or network-observed trusted hash.

Thanks, will do.

> It is also not clear to me that in OFFLINE_COMPLETE catchup, we will actually have streamed the last meta. I don't think we will, given my reading of the code. Not that it matters to users who are passing in --start-ledger and --start-hash, but in this particular test, I think it should be failing when unsafeEmitMeta==false. Maybe double check?

It still passes with `unsafeEmitMeta` forced to `false`, as opposed to using `GENERATE`.  What makes it fail is commenting out that `emitAnyStoredSafeMetaAfterLedgerValidation` (still using iffy name for the moment, will change!) call in `ApplyCheckpointWork::onRun()`:

```cpp
-                lm.emitAnyStoredSafeMetaAfterLedgerValidation();
+                LOG_INFO(DEFAULT_LOG,
+                         "ApplyCheckpointWork::onRun would be calling "
+                         "emitAnyStoredSafeMetaAfterLedgerValidation for "
+                         "ledger {} if it weren't commented out",
+                         lm.getLastClosedLedgerNum());
+                // lm.emitAnyStoredSafeMetaAfterLedgerValidation();
```

If I add another couple log messages, and run the test (which now fails with `emitAnyStoredSafeMetaAfterLedgerValidation` commented out):

```cpp
@@ -509,12 +509,20 @@ LedgerManagerImpl::emitAnyStoredSafeMetaAfterLedgerValidation()
 {
     if (mLedgerCloseMeta)
     {
+        LOG_INFO(DEFAULT_LOG, "Emitting stored metadata for ledger {}",
+                 mLedgerCloseMeta->v0().ledgerHeader.header.ledgerSeq);
         assert(!mApp.getConfig().UNSAFE_EMIT_META);
         assert(mMetaStream);
         mMetaStream->writeOne(*mLedgerCloseMeta);
         mMetaStream->flush();
         mLedgerCloseMeta.reset();
     }
+    else
+    {
+        LOG_INFO(DEFAULT_LOG,
+                 "No stored metadata to emit; last closed ledger {}",
+                 getLastClosedLedgerNum());
+    }
 }
```

```cpp
@@ -594,6 +602,10 @@ LedgerManagerImpl::closeLedger(LedgerCloseData const& ledgerData)
     // LedgerCloseMeta, for streaming to attached clients (typically: horizon).
     if (mMetaStream)
     {
+        LOG_INFO(DEFAULT_LOG,
+                 "closeLedger on ledger {} calling "
+                 "emitAnyStoredSafeMetaAfterLedgerValidation",
+                 header.current().ledgerSeq);
         emitAnyStoredSafeMetaAfterLedgerValidation();
```

Then at the tail of the log I see:

```
2021-03-25T17:56:54.283 <test> [default INFO] closeLedger on ledger 318 calling emitAnyStoredSafeMetaAfterLedgerValidation
2021-03-25T17:56:54.283 <test> [default INFO] Emitting stored metadata for ledger 317
2021-03-25T17:56:54.285 <test> [default INFO] ApplyCheckpointWork::onRun would be calling emitAnyStoredSafeMetaAfterLedgerValidation for ledger 318 if it weren't commented out
2021-03-25T17:56:54.285 <test> [default INFO] closeLedger on ledger 319 calling emitAnyStoredSafeMetaAfterLedgerValidation
2021-03-25T17:56:54.285 <test> [default INFO] Emitting stored metadata for ledger 318
2021-03-25T17:56:54.286 <test> [default INFO] ApplyCheckpointWork::onRun would be calling emitAnyStoredSafeMetaAfterLedgerValidation for ledger 319 if it weren't commented out
2021-03-25T17:56:54.287 <test> [History INFO] Catching up to ledger 319: Download & apply checkpoints: num checkpoints left to apply:0 (100% done)
2021-03-25T17:56:54.287 <test> [History INFO] Catching up to ledger 319: Succeeded: catchup-seq
2021-03-25T17:56:54.287 <test> [History INFO] Catchup finished

[ ... ]

ledger/test/LedgerCloseMetaStreamTests.cpp:236: FAILED:
  REQUIRE( nLcm == 0x13f )
with expansion:
  318 (0x13e) == 319 (0x13f)
```

So the catchup to ledger 319 makes that call (when it's not commented out) for all ledgers up to and including 319.  Not making that call means that the latest meta we've emitted were in `closeLedger` for ledger 319, when we emitted meta for ledger 318.  Here's the log with the real code, with `emitAnyStoredSafeMetaAfterLedgerValidation` restored:

```
2021-03-25T18:01:53.196 <test> [default INFO] closeLedger on ledger 318 calling emitAnyStoredSafeMetaAfterLedgerValidation
2021-03-25T18:01:53.196 <test> [default INFO] No stored metadata to emit; last closed ledger 317
2021-03-25T18:01:53.198 <test> [default INFO] ApplyCheckpointWork::onRun is really calling emitAnyStoredSafeMetaAfterLedgerValidation for ledger 318 (it's not commented out anymore)
2021-03-25T18:01:53.198 <test> [default INFO] Emitting stored metadata for ledger 318
2021-03-25T18:01:53.198 <test> [default INFO] closeLedger on ledger 319 calling emitAnyStoredSafeMetaAfterLedgerValidation
2021-03-25T18:01:53.198 <test> [default INFO] No stored metadata to emit; last closed ledger 318
2021-03-25T18:01:53.200 <test> [default INFO] ApplyCheckpointWork::onRun is really calling emitAnyStoredSafeMetaAfterLedgerValidation for ledger 319 (it's not commented out anymore)
2021-03-25T18:01:53.200 <test> [default INFO] Emitting stored metadata for ledger 319
2021-03-25T18:01:53.201 <test> [History INFO] Catching up to ledger 319: Download & apply checkpoints: num checkpoints left to apply:0 (100% done)
2021-03-25T18:01:53.201 <test> [History INFO] Catching up to ledger 319: Succeeded: catchup-seq
2021-03-25T18:01:53.201 <test> [History INFO] Catchup finished

[ ... ]
All tests passed (8 assertions in 1 test case)
```

So during this test, `ApplyCheckpointWork::onRun` is going through the path that calls `emitAnyStoredSafeMetaAfterLedgerValidation` for each ledger that it closes, including the last.  That causes us still to emit the metadata for the last ledger in this case.  It's never `closeLedger` that emits the meta here, but always `ApplyCheckpointWork`.

Reply:
Hm. Can I rephrase this from my understanding and you can confirm I'm getting the point?

  - When calling `closeLedger` during live externalize events on the network, the sequence of events is:
    - Externalizing ledger N calls `closeLedger`
    - `closeLedger` calls `emitAnyStored...` which flushes the meta that was pending from N-1
    - `closeLedger` buffers the meta for N
    - Externalizing ledger N+1 calls `closeLedger`
    - `closeLedger` calls `emitAnyStored...` which flushes the meta that was pending from N
    - ...
  
  - Whereas when calling `closeLedger` from `ApplyCheckpointWork::onRun`, the sequence of events is:
    - `ApplyCheckpointWork::onRun` N calls `closeLedger`
    - `closeLedger` calls `emitAnyStored...` which has no pending meta to flush
    - `closeLedger` buffers the meta for N
    - `ApplyCheckpointWork::onRun` calls `emitAnyStored...` _again_ which now flushes the meta for N
    - `ApplyCheckpointWork::onRun` N+1 calls `closeLedger`
    - ...
  
Is that correct? If so I .. can see that it works, in the sense that all the replay scenarios get run end-to-end (because we already trust their end hash value) and all the live scenarios are delayed by 1 ledger.

But: it's fairly subtle, involving two different subsystems (live networking and catchup) having a mental model of a state variable that's being set-and-flushed over successive calls to `closeLedger` in `LedgerManager`. It's subtle enough I completely missed this aspect of the design on first glance over the code! 

I would like to request one of two changes to accompany it:

At minimum a comment very clearly spelling out those two different patterns of calling-and-buffering written in the body of `emitAnyStoredSafeMetaAfterLedgerValidation` (or whatever it's renamed to).

More appealing to me would be an enum-typed flag passed to `closeLedger` that indicates the caller's disposition towards meta, eg. `EMIT_META_IMMEDIATELY` or `DELAY_META_BY_ONE_LEDGER` or something, and keep the state variable private to the `LedgerManager` (i.e. remove `emitAnyStoredSafeMetaAfterLedgerValidation` from the public interface of `LedgerManager`). There's a little complexity in here about how you mediate the relationship between that flag and the config flag (eg. whether the caller looks at config or the callee) but I think it keeps the `closeLedger` API from growing a new dimension of statefulness that a caller could corrupt.

Does that make sense? Am I even understanding the situation correctly?

Reply:
> Is that correct? If so I .. can see that it works, in the sense that all the replay scenarios get run end-to-end (because we already trust their end hash value) and all the live scenarios are delayed by 1 ledger.

I think you have those whole sequences exactly right.  As for how to express it, a mental model requiring the subtlety that you cited...

> But: it's fairly subtle, involving two different subsystems (live networking and catchup) having a mental model of a state variable that's being set-and-flushed over successive calls to `closeLedger` in `LedgerManager`. 

...is what I'm hoping to avoid.  Let me see if I can explain, and you can tell me I'm talking nonsense.  :)

I haven't been thinking about this change (which I found shockingly small at first, but have been gradually convincing myself makes sense, though perhaps wrongly!) _operationally_, as in, this happens and then that happens.  That _does_ require coupling between subsystems and a lot of thought about the state at the time that you're in any given code path.

I've been trying to think about it _declaratively_, as a way of satisfying an invariant.  I want to be able to reason about it purely locally, meaning that it should be possible to:

- Look at any given piece of code and immediately know whether it is a place where we should emit the metadata for any particular ledger or not
- Search the whole code base for some single symbol and know that the only places where it could possibly be necessary to emit metadata must be among the places that use that symbol

And I believe that that _is_ possible, and that:

- The symbol you can search for (at least, one of the symbols – there might be others with the same characteristics?) is `LedgerHeaderHistoryEntry::hash`; in particular, this leads us to `LedgerManagerImpl::closeLedger` and `ApplyCheckpointWork::onRun`
- We should emit metadata if and only if the code is confirming that the last closed ledger's hash matches one which has been validated by a hash chain leading back from a `StellarValue` on which the network reached consensus at some point (Graydon greatly simplified this condition for me by explaining that this is what is happening in the `mApplyLedgerSuccess` case of `ApplyCheckpointWork::onRun`)

The "only if" case is the soundness condition:  that our execution of a transaction set has produced a ledger header with a hash that matches one verified by a hash chain leading back from consensus is what tells us that we executed that transaction set correctly.

The "if" case is the completeness condition, and here, I believe, is why:  the condition that we've verified the last closed ledger's hash against a hash chain leading back from consensus is the same condition that we must meet in order to return success from an operation that is trying to catch up to that ledger.  That is what I think ensures that we don't have off-by-one "missed one ledger's meta" errors here:  if a catchup operation ever returned success from a request to catch up to ledger `N` without having checked the hash resulting from closing ledger `N` against a hash chain, then the catchup operation itself is implemented wrongly.  Therefore, if we emit meta at the place in the code where we've just done that check (and it has passed!), then we will always have emitted meta for ledger `N` before catchup gets a chance to tell its caller that it has successfully caught up to ledger `N`.

Broadening the model beyond the question of when to emit meta, I am thinking of the root cause of the problem that we emit incorrect meta after a one-ledger fork as the existing code's conflating two conditions which are and must be separate:  consensus on the _contents_ of a ledger (represented by a `StellarValue`) and consensus on the _results_ of applying that ledger (of which the meta are one representation, a particularly verbose one for the benefit of Horizon).  It is inevitable that those conditions must be met at different times, because nodes can't even start computing the results of applying a ledger until consensus has been reached on the contents of that ledger.  By emitting meta as soon as we've applied a ledger, we're treating the results as a deterministic, pure function of the contents.  The case of a one-ledger fork caused by a local corruption is the case where that does not hold.

One solution, therefore, would be to have another round of consensus on the _results_ of applying each ledger.  If that phase had short timeouts, it might be the quickest way of reaching a point where we could safely emit meta.  But it would increase network traffic, and also it would take me a hundred years and a thousand bugs to implement.  Fortunately, the `StellarValue` for ledger `N+1` certifies that consensus has been reached not only on the contents of ledger `N+1`, but also on the _results_ of applying ledger `N`, and, via the chain formed by the headers, the results of applying each ledger in `[1..N]`.  That's why the one-ledger-delay implementation provides safety.

So I might have this all wrong, but my conception of this problem, and the proposed solution, is that we're switching to emitting meta for ledger `N` when we reach consensus on the _results_ of applying ledger `N`, instead of when we've only reached consensus on the _contents_ of ledger `N`, as currently written.  That we do so by (when we're in sync with a network) implicitly recognizing that consensus on the contents of ledger `N+1` implies consensus on the results of applying ledger `N` is an implementation detail; we could achieve the same with an explicit consensus round on the results.  But, if I really can get away with it, I would much rather do so just by factoring out one eight-line function and calling it in one new place.  ;)

If I _don't_ have it all wrong and the above is actually true, then I think the next question is, what names/APIs would best express it?

Reply:
I can see what you're saying, and I think mostly agree with the clarification to the mental model of consensus you're applying here (though it's maybe worth mentioning that not all forks are caused by local corruption / nondeterminism -- they can also be caused by quorum misconfiguration).

My concern is less to do with mental model of consensus and more to do with more basic code modularity:
  - Callers need to think about a new implicit state variable in `LedgerManager`.
  - The two different types of caller require two different patterns of call (close, close, close vs. close, flush, close, flush).
  - If callers call the flush function at an unexpected time or in the wrong order, they corrupt the state variable.

I'm asking if it's possible to localize the maintenance of _the state variable_ to `LedgerManager`, so callers don't have to know about it / call things in the right order to maintain its invariants. I'd like to keep its invariants private to `LedgerManager` if we can.

Following your comments on mental model, I think this can even be extended to callers lacking awareness of the concept of metadata flushing altogether: maybe the caller only needs to express whether they're providing a ledger that's had its hash confirmed by a subsequent consensus ledger or not. That leaves it up to the `LedgerManager` to interpret that as a sign to flush or delay the metadata emission, and (as in the current code) the `LedgerManager` is the only module that really has any idea there's a metadata stream at all.

Does that make sense?

Reply:
> I think this can even be extended to callers lacking awareness of the concept of metadata flushing altogether: maybe the caller only needs to express whether they're providing a ledger that's had its hash (and txSetResultHash) confirmed by a subsequent consensus ledger or not. That leaves it up to the LedgerManager to interpret that as a sign to flush or delay the metadata emission, and (as in the current code) the LedgerManager is the only module that really has any idea there's a metadata stream at all.

> Does that make sense?

I think so!  Is your idea that `LedgerCloseData` could be given something like a `std::optional<Hash>` which, if it has a value, represents the expected `txResultsHash`, and then `closeLedger` itself could (without callers knowing that this was exactly what it was doing with the information) emit metadata right away if there is an expected `txResultsHash` and if it matches at the end of `closeLedger`, abort right away if there is an expected `txResultsHash` and it _doesn't_ match, and defer emitting metadata if there is no expected `txResultsHash`?  And then `ApplyCheckpointWork` would provide an expected `txResultsHash` from the archive, whereas `valueExternalized` would not provide one?

Reply:
Maybe? I mean, again, the `txSetResultHash` matching only traps forks caused by local corruption / nondeterminism, whereas the set of things-that-diverge due to a quorum misconfiguration is a bit wider than that (the other fields in the ledger header -- `baseFee`, `maxTxSetSize`, `ledgerVersion` and so forth). But it might be that we treat passing an expected `txSetResultHash` as implying "we confirmed everything about this ledger, including this `txSetResultHash`, in a subsequent ledger".

If it's too fussy to fish out the `txSetResultHash` from the existing code when doing replay (which I think it might be), I don't think it'd necessarily be any worse to just pass a 2-state flag indicating whether the caller does or doesn't have later-confirmation-confidence in the ledger being closed. Passing a `txSetResultHash` does give a slightly-earlier abort _during_ close if there's a mismatch, but absent that any mismatch from nondeterminism will trap on the next ledger anyway (which will happen immediately during replay), so I don't know that it's worth too much contortion to catch early.

Reply:
> Maybe? I mean, again, the txSetResultHash matching only traps forks caused by local corruption / nondeterminism, whereas the set of things-that-can-diverge due to a quorum misconfiguration is a bit wider than that (the other fields in the ledger header -- baseFee, maxTxSetSize, ledgerVersion and so forth). But it might be that we treat passing an expected txSetResultHash as implying "we confirmed everything about this ledger, including this txSetResultHash, in a subsequent ledger".

Right, thanks -- I should have said the expected ledger header hash; I think that should detect non-determinism from any source -- is that true?

> If it's too fussy to fish out the `txSetResultHash` from the existing code when doing replay (which I think it might be), I don't think it'd necessarily be any worse to just pass a 2-state flag indicating whether the caller does or doesn't have later-confirmation-confidence in the ledger being closed. Passing a `txSetResultHash` does give a slightly-earlier abort _during_ close if there's a mismatch, but absent that any mismatch from nondeterminism will trap on the next ledger anyway (which will happen immediately during replay), so I don't know that it's worth too much contortion to catch early.

The reason I thought it'd have to be a `std::optional<Hash>` is that I'm not following how `LedgerManager` would be able to determine from a flag alone whether it should emit metadata for the ledger that it's closing -- if it knows that the caller knows what the results of closing the ledger should be, but doesn't know what those results _are_, then if it did choose to emit metadata, then even though incorrect results would trap as soon as `closeLedger` returned, wouldn't it already be too late?

Reply:
> if it knows that the caller knows what the results of closing the ledger should be, but doesn't know what those results are, then if it did choose to emit metadata, then even though incorrect results would trap as soon as closeLedger returned, wouldn't it already be too late?

Oh, yeah, I guess so -- fair point! I hadn't noticed that the flush in the `onRun` caller happens after an expected-hash-check by the caller, thought it was also going to delay noticing until next iteration anyway.

Reply:
Okay, I really like the idea of callers giving `closeLedger` an optional expected ledger hash; it makes it explicit that there are two different things we might have consensus on already:  the contents (represented by a `StellarValue`), which we _definitely_ have consensus on or we wouldn't be calling `closeLedger`, and the results (represented by the value that, if present, we expect the ledger hash to have after the close is complete), which we might or might not have consensus on already, depending on the context in which we're being called.  `closeLedger` becomes "externalize contents of ledger `N`, and _maybe_ externalize results of ledger N".  Also, `closeLedger(N)` always implies "externalize results of ledgers `[1..N-1]`".  And what `LedgerManagerImpl` privately chooses to do with "externalize contents" is apply transactions and upgrades; what it does with "externalize results" is stream meta.

My first thought was to stick an `optional<Hash>` into `LedgerCloseData`; @MonsieurNicolas 's was to provide an overload of `closeLedger` with a `Hash` parameter.  Is there a reason to prefer one or the other clearly?